core/arch/arm/plat-k3/drivers/sec_proxy.c

```diff
@@ -0,0 +1,294 @@
+	if (ret) {
+		EMSG("Thread %s verification failed. ret = %d\n", spt->name,
+		     ret);
+		return ret;
+	}
+
+	/* Check the message size. */
+	if (msg->len > SEC_PROXY_MAX_MSG_SIZE) {
+		EMSG("Thread %s message length %zu > max msg size %d\n",
+		     spt->name, msg->len, SEC_PROXY_MAX_MSG_SIZE);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	/* Send the message */
+	data_reg = spt->data + SEC_PROXY_DATA_START_OFFS;
+	num_words = msg->len / sizeof(uint32_t);
+	for (i = 0; i < num_words; i++) {
+		memcpy(&data_word, &msg->buf[i * 4], sizeof(uint32_t));
+		io_write32(data_reg, data_word);
+		data_reg += sizeof(uint32_t);
+	}
+
+	trail_bytes = msg->len % sizeof(uint32_t);
+	if (trail_bytes) {
+		uint32_t data_trail = 0;
+
+		i = msg->len - trail_bytes;
+		while (trail_bytes--) {
+			data_trail <<= 8;
+			data_trail |= msg->buf[i++];
+		}
+
+		io_write32(data_reg, data_trail);
+		data_reg += sizeof(uint32_t);
+	}
+
+	/*
+	 * 'data_reg' indicates next register to write. If we did not already
+	 * write on tx complete reg(last reg), we must do so for transmit
+	 */
+	if (data_reg <= (spt->data + SEC_PROXY_DATA_END_OFFS))
+		io_write32(spt->data + SEC_PROXY_DATA_END_OFFS, 0);
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * k3_sec_proxy_recv() - Receive data from a Secure Proxy thread
+ * @msg: Pointer to k3_sec_proxy_msg
+ *
+ * Return: 0 if all goes well, else appropriate error message
+ */
+TEE_Result k3_sec_proxy_recv(struct k3_sec_proxy_msg *msg)
+{
+	struct k3_sec_proxy_thread *spt = &spts[SEC_PROXY_RX_THREAD];
+	int num_words = 0, i = 0;
+	int trail_bytes = 0;
+	uint32_t data_trail = 0;
+	uint32_t data_word = 0;
+	uintptr_t data_reg = 0;
+	TEE_Result ret = TEE_SUCCESS;
+
+	FMSG("Verifying thread\n");
+	ret = k3_sec_proxy_verify_thread(THREAD_DIR_RX);
+	if (ret) {
+		EMSG("Thread %s verification failed. ret = %d\n", spt->name,
+		     ret);
+		return ret;
+	}
+
+	/* Receive the message */
+	data_reg = spt->data + SEC_PROXY_DATA_START_OFFS;
+	num_words = msg->len / sizeof(uint32_t);
+	for (i = 0; i < num_words; i++) {
+		data_word = io_read32(data_reg);
+		memcpy(&msg->buf[i * 4], &data_word, sizeof(uint32_t));
+		data_reg += sizeof(uint32_t);
+	}
+
+	trail_bytes = msg->len % sizeof(uint32_t);
+	if (trail_bytes) {
+		data_trail = io_read32(data_reg);
+		data_reg += sizeof(uint32_t);
+
+		i = msg->len - trail_bytes;
+		while (trail_bytes--) {
+			msg->buf[i] = data_trail & 0xff;
+			data_trail >>= 8;
+		}
+	}
+
+	/*
+	 * 'data_reg' indicates next register to read. If we did not already
+	 * read on rx complete reg(last reg), we must do so for receive
+	 */
+	if (data_reg <= (spt->data + SEC_PROXY_DATA_END_OFFS))
+		io_read32(spt->data + SEC_PROXY_DATA_END_OFFS);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result k3_sec_proxy_init(void)
+{
+	uint32_t target_data = 0, cfg_scfg = 0, cfg_rt = 0;
+	struct k3_sec_proxy_thread *thread;
+	int rx_thread = SEC_PROXY_RESPONSE_THREAD;
+	int tx_thread = SEC_PROXY_REQUEST_THREAD;
+
+	FMSG("Initialize %s\n", __func__);
```

Comment:
Doesn't `FMSG()` print the function name by default already? In addition you have a `DMSG()` immediately after so this seems redundant.

Reply:
ah yes, makes sense. Will remove the `FMSG`